plugins/sharding/logical_enumerator.rb

```diff
@@ -83,16 +85,29 @@ def each_shard_with_around(order)
           end
 
           shard = Shard.new(name, @shard_key_name, shard_range_data)
+          previous_shard = shards.last
+          if previous_shard
+            shard.previous_shard = previous_shard
+            previous_shard.next_shard = shard
+          end
           shards << shard
           @shards << shard
           next if shards.size < 3
           yield(*shards)
-          shards.shift
+          shifted_shard = shards.shift
+          next unless unref_immediately
+          next if shifted_shard.nil?
+          shifted_shard.unref
+          @shards.shift
         end
 
         if shards.size == 2
           yield(shards[0], shards[1], nil)
         end
+        if unref_immediately
+          @shards.each(&:unref)
+          @shards.clear
```

Comment:
I have changed to use `#unref`.
Additionaly, I have added `@shard.clear` to `LogicalEnumerator.unref` in order to ensure executing `Shard.unref` only once even if `LogicalEnumerator.unref` is called multiple times.

Also, I have changed to ensure executing this `unref` because a caller of this function sometimes break a loop before reaching the end of this function.